internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,357 @@
+/*
+Copyright 2021 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"context"
+	"errors"
+	"strconv"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+	"github.com/kube-storage/spec/lib/go/replication"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+//  imageMirroringMode is used to indicate the mirroring mode for an RBD image.
+type imageMirroringMode string
+
+const (
+	// imageMirrorModeSnapshot uses snapshots to propagate RBD images between
+	// ceph clusters.
+	imageMirrorModeSnapshot imageMirroringMode = "snapshot"
+)
+
+// imageMirroringState represents the image mirroring state.
+type imageMirroringState string
+
+const (
+	// If the state is up+replaying, then mirroring is functioning properly.
+	// up means the rbd-mirror daemon is running, and replaying means
+	// this image is the target for replication from another storage cluster.
+	upAndReplaying imageMirroringState = "up+replaying"
```

Comment:
Nit: the goal should definitely be to not rely on CLI output for determining status when the librbd API provides this as decomposed fields.

Reply:
Yes the PR is in progress in go-ceph will switch from CLI to go-ceph once its available